### PR TITLE
Composer

### PR DESCRIPTION
**Pull request by generalredneck**
This is a "pre-abstraction" level of build...

Eventually we would like to make it so that the build system could be controlled by extensions so they can provide their own commands to run as part of the build.

Additionally, allowing a build plugin to hook into create would allow one to set up a default project with the neccessary files to make a build work. Composer is an example of this type of scenario. Drush is too actually. 

Drush requires a drush.make.yml file, composer requires a composer.json file and a couple of supporting scripts.

This builds composer support for a modified version of drupal-composer/drupal-project. I've only tested this with Drupal 8. I've not successfully tested this build with a core like pantheon/drops-8. (that particular one was very difficult to write a composer.json for because they commit the vendor folder)

There are some scaffolding that drupal-composer/drupal-scaffold builds that I had to remove so that aquifer wouldn't fail. This is part of the afterBuild callback.

The steps to get a build working is something of the following:
- run `aquifer create test`
- modify aquifer.json to have
  
  ```
  "build": {
    "method": "composer",
    "directory": "build",
    "makeFile": "composer.json"
  } 
  ```
- download [drupal-composer/drupal-project: 8.x](https://github.com/drupal-composer/drupal-project/archive/8.x.zip)
- copy composer.json and the scripts directory into your test project
- change all instances of "web" in both composer.json and scripts/composer/ScriptHandler.php to whatever `build.directory` is in aquifer.json. In this case "build".
- Add the following to composer.json 
  
  ```
    "config": {
        "vendor-dir": "build/vendor"
    }
  ```

This will get you a basic build

### CODE REVIEW

File: lib/build.api.js [hunk 1]
```diff
@@ -10,6 +10,8 @@ const _ = require('lodash');
 const fs = require('fs-extra');
 const del = require('del');
 const path = require('path');
+const glob = require("glob");
+const fx = require('mkdir-recursive');
```
Comment (patrickocoffeyo) [inline lib/build.api.js]: No need for this, fs-extra mkdirSync provides recursive mkdirs: https://github.com/jprichardson/node-fs-extra#mkdirsdir-callback

File: lib/build.api.js [hunk 2]
```diff
@@ -10,6 +10,8 @@ const _ = require('lodash');
 const fs = require('fs-extra');
 const del = require('del');
 const path = require('path');
+const glob = require("glob");
```
Comment (patrickocoffeyo) [inline lib/build.api.js]: Can we use single ticks instead of quotes, for consistency?

File: lib/build.api.js [hunk 3]
```diff
@@ -210,15 +235,15 @@ class Build {
...
             this.aquifer.console.log('Destination directory does not exist. Creating: ' + destBase, 'status');
-            fs.mkdirSync(destBase);
+            fs.mkdirsSync(destBase);
           }
 
           // If symlinking is turned on, symlink custom files. Else, copy.
           if (this.options.symlink) {
 
             // Make sure the destination base path exists.
             if (!fs.existsSync(destBase)) {
```
Comment (JeffTomlinson) [inline lib/build.api.js:245]: This legacy block is not needed. Already took care of it above.

**Comment by thebruce** [comment]
Functional works here. Aquifer build does all the composer things!

Note for testing: Preserve the file structure of scripts/composer/ScriptHandler.php

That's probably obvious to anybody, but i messed up on the first go.

File: lib/build.api.js [hunk 4]
```diff
@@ -210,15 +235,15 @@ class Build {
           // Make sure the destination base path exists.
           if (!fs.existsSync(destBase)) {
             this.aquifer.console.log('Destination directory does not exist. Creating: ' + destBase, 'status');
-            fs.mkdirSync(destBase);
+            fs.mkdirsSync(destBase);
```
Comment (mattgrill) [inline lib/build.api.js:238]: Should probably leave a note that this is from `fs-extra`, https://github.com/jprichardson/node-fs-extra#mkdirsdir-callback

File: lib/build.command.js [hunk 5]
```diff
@@ -20,15 +20,13 @@ module.exports = function (Aquifer) {
...
     let copy = options.copy === true;
-    let directory = options.directory || settings.paths.build;
+    let directory = options.directory || settings.build.directory;
     let build = new Aquifer.api.build(Aquifer);
 
     // Create build.
     build.create(directory, {
-      symlink: !copy,
-      makeFile: makeFile
+      symlink: !copy
```
Comment (mattgrill) [inline lib/build.command.js:29]: Not usually a fan of storing things in yoda-like variables.

File: package.json [hunk 6]
```diff
@@ -39,6 +39,7 @@
     "commander": "^2.7.1",
     "del": "^1.2.1",
     "fs-extra": "^0.18.0",
+    "glob": "^7.0",
```
Comment (mattgrill) [inline package.json:42]: @chasingmaxwell What happened with npm shinkrwrap?

File: lib/build.command.js (hunk 5 above)
Comment (generalredneck) [inline lib/build.command.js:29]: Yoda-like?More detail please if there is a problem as its not apparent the expected results from your review here.

**Pull request merged**